### PR TITLE
OOP school library: set up associations

### PR DESCRIPTION
**Pull request by DebyGrey**
In this PR, I

- Created a class Classroom with the following:
  - @ label instance variable, should be initialized in the constructor.
  - Setter and getter for @ label (remember about attr_accessor).
- Created the has-many/belongs-to relationship between Classroom and Student. The following should be implemented:
  - Created the has-many side (a classroom has many students).
  - Created the belongs-to side (a student belongs to a classroom).
  - Ensured that when adding a student to a classroom it also sets the classroom for the student.
  - Ensured that when setting the classroom for a student it also adds it to the classrooms' students.
- Created a class Book with the following:
  - @ title and @ author instance variables, should be initialized in the constructor.
  - Setters and getters for instance variables (remember about attr_accessor).
- Created a class Rental with the following:
  - @ date instance variable, should be initialized in the constructor.
  - Setter and getter for @ date (remember about attr_accessor).
- Created the many-to-many (also has-many-through) relationship between Person and Book using the intermediate class Rental. The following should be implemented:
  - Created the has-many side of Book and Rental (a book has many rentals).
  - Created the belongs-to side of Rental and Book (a rental belongs to a book).
  - Created the has-many side of Person and Rental (a person has many rentals).
  - Created the belongs-to side of Rental and Person (a rental belongs to a person).
  - Modified the constructor of Rental so Book and Person are set in it.

### CODE REVIEW

File: classroom.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
+class Classroom
+  attr_accessor :label, :students
```
Comment (mikethreels) [inline classroom.rb]: - [x] for the student it will be enough to only set up the attribute reader.

File: student.rb [hunk 2]
```diff
@@ -1,6 +1,8 @@
 class Student < Person
+  attr_accessor :classroom
+
   def initialize(classroom, age, name: 'unknown', parent_permission: true)
-    super(age, name, parent_permission,)
+    super(age, name, parent_permission)
     @classroom = classroom
   end
```
Comment (mikethreels) [inline student.rb:5]: - [x] there should be a method that serves as the 'belongs-to' side (a student belongs to a classroom). have a look at [the article](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_relationships_by_examples.md#both-ways-relationships) with examples

File: book.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
+class Book
+  attr_accessor :title, :author
```
Comment (mikethreels) [inline book.rb]: - [x] for the association to work Getters and Setters for @rentals should be present

File: book.rb [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+  def initialize(title, author)
+    @title = title
+    @author = author
+    @rentals = []
+  end
+
+  def add_rental(rental)
+    @rentals << rental
+  end
```
Comment (mikethreels) [inline book.rb]: - [x] the add_rental method should look a bit different in order for this to work correctly have a look at the example in a previous lesson: https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_relationships_by_examples.md#many-to-many-relationship

File: rental.rb [hunk 5]
```diff
@@ -0,0 +1,12 @@
+class Rental
+  attr_accessor :date, :person, :book
+
+  def initialize(date, person, book)
+    @date = date
+    @person = person
+    @book = book
+
+    person.add_rental(self)
+    book.add_rental(self)
```
Comment (mikethreels) [inline rental.rb]: - [ ] the rental should add itself to the book and the person by being pushed to book.rentals and person.rentals not by calling the method.

File: person.rb [hunk 6]
```diff
@@ -4,20 +4,25 @@
...
+    @rentals = []
   end
 
   def correct_name
     @name
   end
 
+  def add_rental(rental)
+    @rentals << rental
+  end
```
Comment (mikethreels) [inline person.rb]: - [ ] the add_rental method should look a bit different in order for this to work correctly have a look at the example in a previous lesson: https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_relationships_by_examples.md#many-to-many-relationship

**Review by mikethreels — CHANGES_REQUESTED** [review]
Hi Deborah, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
- good job starting with the associations.

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: student.rb [hunk 7]
```diff
@@ -1,7 +1,10 @@
 class Student < Person
+  attr_accessor :classroom
+
   def initialize(classroom, age, name: 'unknown', parent_permission: true)
-    super(age, name, parent_permission,)
+    super(age, name, parent_permission)
     @classroom = classroom
+    classroom&.add_student(self)
   end
```
Comment (tufoinnkuo10) [inline student.rb]: - [x] Great job so far. like the previous reviewer noted, there should be a clear method that serves as the 'belongs-to' side (a student belongs to a classroom).  something like

```ruby
def classroom=(classroom)
      @classroom = classroom
      classroom.students.push(self) unless classroom.students.include?(self)
  end
```

File: rental.rb (hunk 5 above)
Comment (tufoinnkuo10) [inline rental.rb]: - [ ] nice work. like the previous reviewer noted, the rental should add itself to the book and the person by being pushed to book.rentals and person.rentals `achieving the 'belongs-to' association` eg 

```ruby
    @book = book
    book.rentals << self

    @person = person
    person.rentals << self
```

File: person.rb [hunk 8]
```diff
@@ -4,20 +4,26 @@
...
   end
 
   def correct_name
     @name
   end
 
+  def add_rental(rental)
+    @rentals << rental
+    rental.person = self
+  end
```
Comment (tufoinnkuo10) [inline person.rb]: - [ ] like the previous reviewer noted, The person itself should be sent as a parameter to create the new rental achieving the 'has-many' association

```ruby
def add_rental(book, date)
    Rental.new(date, book, self)
```

File: book.rb [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+
+  def initialize(title, author)
+    @title = title
+    @author = author
+    @rentals = []
+  end
+
+  def add_rental(rental)
+    @rentals << rental
+    rental.book = self
```
Comment (tufoinnkuo10) [inline book.rb]: - [x] nice work adding the rental method. you know it includes the person, the book and rental date. please note that the book itself should be sent as a parameter to create the new rental achieving the 'has-many' association.` you should have something like

```ruby
def add_rental(person, date)
    Rental.new(date, self, person)
  end
```

**Review by tufoinnkuo10 — CHANGES_REQUESTED** [review]
Hi @DebyGrey 

There are some issues that you still need to work on to go to the next project but you are almost there!

To Highlight
✅ You used the correct Gitflow
✅ Descriptive pull request message
✅ linter error checks validated
✅ Documented your work professionally
✅ associations created

Required Changes ♻️
Check the comments under the review
Optional suggestions
_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better.

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise.
As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using this [form.](https://airtable.com/shrQAqnBwek5a0O0s)

**Review by tufoinnkuo10 — APPROVED** [review]
## Status: Approved :heavy_check_mark:

Hi @DebyGrey 

Your project is complete! Great job for implementing the changes🥇. There is nothing else to say other than... it's time to merge it 🍾🚢 :shipit: . Congratulations!  🎉💯🌟 

## Good Points 👍
- Descriptive Pull Request
- Associations well created
- methods added
- No linter errors

<br>
Cheers..Happy coding!..and keep soaring higher! 💻 🍷 🚀 

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have 2 more limited reviews per this project. If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**